sys/drv/pit.c

```diff
@@ -11,56 +11,75 @@ typedef struct pit_state {
+  /* values since last counter read */
+  uint16_t prev_ticks16; /* number of ticks */
+  /* values since initialization */
+  uint32_t ticks; /* number of ticks modulo TIMER_FREQ*/
+  uint32_t sec;   /* seconds */
 } pit_state_t;
 
 #define inb(addr) bus_read_1(pit->regs, (addr))
 #define outb(addr, val) bus_write_1(pit->regs, (addr), (val))
 
-static void pit_set_frequency(pit_state_t *pit) {
+static inline void pit_set_frequency(pit_state_t *pit) {
   outb(TIMER_MODE, TIMER_SEL0 | TIMER_16BIT | TIMER_RATEGEN);
-  outb(TIMER_CNTR0, pit->period_cntr & 0xff);
-  outb(TIMER_CNTR0, pit->period_cntr >> 8);
+  outb(TIMER_CNTR0, pit->period_ticks & 0xff);
+  outb(TIMER_CNTR0, pit->period_ticks >> 8);
 }
 
-static uint16_t pit_get_counter16(pit_state_t *pit) {
+static inline uint16_t pit_get_counter(pit_state_t *pit) {
   uint16_t count = 0;
   outb(TIMER_MODE, TIMER_SEL0 | TIMER_LATCH);
   count |= inb(TIMER_CNTR0);
   count |= inb(TIMER_CNTR0) << 8;
-  return count;
+
+  /* PIT counter counts from n to 1, we make it ascending from 0 to n-1*/
+  return pit->period_ticks - count;
 }
 
-static uint64_t pit_get_counter64(pit_state_t *pit) {
-  SCOPED_INTR_DISABLED();
-  uint16_t cntr16_now = pit_get_counter16(pit);
-  /* PIT counter counts from n to 1 and when we get to 1 an interrupt
-     is send and the counter starts from the beginning (n = pit->period_cntr).
-     We do not guarantee that we will not miss the whole period (n ticks) */
-  uint16_t ticks = pit->cntr16_prev_read - cntr16_now;
-  if (pit->cntr16_prev_read < cntr16_now)
-    ticks += pit->period_cntr;
+static void pit_update_time(pit_state_t *pit) {
+  assert(intr_disabled());
+  uint32_t last_ticks = pit->ticks;
+  uint32_t last_sec = pit->sec;
+  uint16_t now_ticks16 = pit_get_counter(pit);
+  uint16_t ticks_passed = now_ticks16 - pit->prev_ticks16;
+  if (pit->prev_ticks16 > now_ticks16) {
+    pit->noticed_overflow = true;
+    ticks_passed += pit->period_ticks;
+  }
 
   /* We want to keep the last read counter value to detect possible future
    * overflows of our counter */
-  pit->cntr16_prev_read = cntr16_now;
-
-  uint32_t oldlow = pit->cntr64.lo;
-  pit->cntr64.lo += ticks;
-  if (oldlow > oldlow + ticks)
-    pit->cntr64.hi++;
-
-  return pit->cntr64.val;
+  pit->prev_ticks16 = now_ticks16;
+
+  pit->ticks += ticks_passed;
+  if (pit->ticks >= TIMER_FREQ) {
+    pit->ticks -= TIMER_FREQ;
+    pit->sec++;
+  }
+  assert(last_sec < pit->sec ||
+         (last_sec == pit->sec && last_ticks < pit->ticks));
+  assert(pit->ticks < TIMER_FREQ);
 }
 
 static intr_filter_t pit_intr(void *data) {
   pit_state_t *pit = data;
 
-  /* XXX: It's still possible for a tick to be lost. */
-  (void)pit_get_counter64(pit);
+  /* XXX: It's still possible for periods to be lost. */
+  pit_update_time(pit);
+  if (!pit->noticed_overflow) {
+    pit->ticks += pit->period_ticks;
+    if (pit->ticks >= TIMER_FREQ) {
+      pit->ticks -= TIMER_FREQ;
+      pit->sec++;
+    }
```

Comment:
This duplicates some logic from `pit_update_time()`, can you factor it out into another function, e.g. `pit_incr_ticks(uint32_t ticks)`?